### ISSUE TITLE
Implement helpers for SNIP-21, SNIP-22, and SNIP-23 interfaces

### ISSUE THREAD

**Issue by reuvenpo** [enigmampc/secret-toolkit#20]
Extend the methods available under `secret_toolkit::snip20` to also include the new interface specified in:
[SNIP-21](https://github.com/SecretFoundation/SNIPs/blob/master/SNIP-21.md), [SNIP-22](https://github.com/SecretFoundation/SNIPs/blob/master/SNIP-22.md), and [SNIP-23](https://github.com/SecretFoundation/SNIPs/blob/SNIP-23/SNIP-23.md).

Alternatively, implement new modules called `snip21`, `snip22`, and `snip23` which include versions of the modified interfaces that are aware of the new features, as well as support for the new message types.

**Comment by floAr** [comment]
A quick question to this: Would the new modules only include the additions and changes to snip20 or would you rather have it to mirror everything? 

Concrete example: If I call *Transfer* on a snip21 do I import it from ::snip20 or ::snip21 ?

**Comment by reuvenpo** [comment]
Good question, that's the discussion that i'd like to have here.

We could perhaps provide an alternative implementations to some things under `snip21`, e.g. a `Transfer` implementation with the `memo`, and then `pub use` (reexport) the interfaces that didn't change. That way we don't have to implement things twice, but users can still access everything they need from one module.

The question though then becomes: does `snip22` include `snip21`? Since `snip22` is completely orthogonal to `snip20` and `snip21` we could have it export only the new messages that it defines, but then there's a certain asymmetry between them. Is that ok? I think it's a fine decision to make for now.

(I'm gonna explain that in the relevant PRs in a few minutes but) We're going to shelve `snip23` (allowance improvements) for now, but if we were to include it, how would we represent it in the toolkit? just like `snip21`, it adds new messages AND extends existing ones. What should/would the `snip23` module export?

**Comment by floAr** [comment]
I would personally prefer asymmetry in the packages. So the functionally resides in the package it is defined in and I compose it in my contracts. That will mean that I potentially have multiple crates to import but it also makes it very clear which functionality I am using when writing the contract. 
For something like the `snip23` it would make sense to reexport functions and then I as a developer can choose which one I want to import and use. Here I am not sure how well intellisense works for Rust but in C# it would give me the tooltip like 'You are missing function Transfer, do you want to import it from ::snip20, ::snip22 or snip:33

**Comment by reuvenpo** [comment]
Yeah, at least in CLion if you use a symbol you didn't import, it offers several options, and pastes in the appropriate import line next to your other imports.

**Comment by reuvenpo** [comment]
So just to make sure, in your opinion `snip21` should or shouldn't export all the symbols from `snip20`?

**Comment by floAr** [comment]
I think it should **not** export unchanged symbols. This is based on my following understanding (and please correct me if I am wrong here): _Snip2X implementations are composable functionality- So my token could be based on snip20 and implement snip22 and 21 as well. Or potentially only 22 and so forth._ 
If we re-export snip20 from snip22 I would expect to have similar structure for snip21 and a similar one for snip21+ snip22 so the number of permutations rises exponentially with future improvements.
Instead I would suggest that I have a base functionality in snip20 and I can choose to pick a modern replacement from another crate to use it instead. This does still not solve the problem how I would solve the problem if two snips improve on the same symbol, but for the current moment this seems to be the clearest way for me.
